### PR TITLE
Added "constructor" module and ST_MAKEENVELOPE function

### PR DESCRIPTION
**Pull request by alex-graciano**
Added the requested function in 
https://app.clubhouse.io/cartoteam/story/148714/add-st-makeenvelope

The BigQuery ST_MAKEPOLYGON function always returns a counterclockwise polygon. For these both inputs:
`ST_MAKEPOLYGON(ST_GeogFromText('LINESTRING(10 10,10 11,11 11, 11 10)')) -- counterclockwise` 
`ST_MAKEPOLYGON(ST_GeogFromText('LINESTRING(11 10, 11 11, 10 11, 10 10)')) -- clockwise`

The same output is generated:

`POLYGON((11 10, 11 11, 10 11, 10 10, 11 10))`

Is it important to our implementation? Is there something I am missing?

### CODE REVIEW

**Comment by jgoizueta** [comment]
Regarding polygon orientation: https://cloud.google.com/bigquery/docs/gis-data?hl=in#polygon_orientation

**Comment by alex-graciano** [comment]
> Regarding polygon orientation: https://cloud.google.com/bigquery/docs/gis-data?hl=in#polygon_orientation

Exactly, I am reading this. I am editing the PR while @Jesus89 investigate about the CLA error.

File: constructors/bq/sql/ST_MAKEENVELOPE.sql
Comment (jgoizueta) [inline constructors/bq/sql/ST_MAKEENVELOPE.sql:12]: This might be a micro-optimization, but we could put the vertices in counter-clockwise order and then use `ST_MAKEPOLYGONORIENTED` to save the check and reversal that `ST_MAKEPOLYGON` would do.

Comment (jgoizueta) [reply constructors/bq/sql/ST_MAKEENVELOPE.sql:12]: 🤔 but then it would be safer not to trust the caller and use `GREATEST(xmin, xmax)`, `LEAST(xmin, xmax)` etc.

File: constructors/library/constructors/constructors_version.js
Comment (jgoizueta) [inline constructors/library/constructors/constructors_version.js:9]: Since you're adding a JS lib with a version function too I think you should use it here. (this is the kind of thing that @Jesus89 code generation tool will save us from)

File: constructors/bq/sql/ST_MAKEENVELOPE.sql
Comment (alex-graciano) [reply constructors/bq/sql/ST_MAKEENVELOPE.sql:12]: I am keeping as simple as PostGIS does (I think they do not make any check). But I could add them without any problem

File: tools/setool/README.md
Comment (jgoizueta) [inline tools/setool/README.md:1]: Hey, why is this included in this PR?

File: constructors/library/constructors/constructors_version.js
Comment (alex-graciano) [reply constructors/library/constructors/constructors_version.js:9]: Yes, this is autogenerated code from the setool

File: constructors/bq/sql/ST_MAKEENVELOPE.sql
Comment (jgoizueta) [reply constructors/bq/sql/ST_MAKEENVELOPE.sql:12]: On second thought, maybe we should not secong-guess the user. What if xmin= 179, xmax = -181. By the way, we should check what does ST_MAKEPOLYGON does with this kind of input.

File: tools/setool/README.md
Comment (alex-graciano) [reply tools/setool/README.md:1]: My bad, I merged previously this on my branch. This is actually reviewed in https://github.com/CartoDB/carto-spatial-extension/pull/73/files

File: constructors/bq/sql/ST_MAKEENVELOPE.sql
Comment (jgoizueta) [reply constructors/bq/sql/ST_MAKEENVELOPE.sql:12]:
```sql
ST_MAKEPOLYGON(ST_GeogFromText('LINESTRING(179 10,179 11,-179 11, -179 10)'))
-- 	 POLYGON((-179 10, -179 11, 179 11, 179 10, -179 10))
ST_MAKEPOLYGON(ST_GeogFromText('LINESTRING(-179 10, -179 11, 179 11, 179 10)'))
-- POLYGON((-179 10, -179 11, 179 11, 179 10, -179 10))
ST_MAKEPOLYGONORIENTED([ST_GeogFromText('LINESTRING(179 10,179 11,-179 11, -179 10)')])
-- POLYGON((179 10, 179 11, -179 11, -179 10, 179 10))
ST_MAKEPOLYGONORIENTED([ST_GeogFromText('LINESTRING(-179 10, -179 11, 179 11, 179 10)')])
-- POLYGON((-179 10, -179 11, 179 11, 179 10, -179 10))
```
So, `ST_MAKEPOLYGON` handles the change of date line well, and `ST_MAKEPOLYGONORIENTED` maintains the input order

File: constructors/library/constructors/constructors_version.js
Comment (jgoizueta) [reply constructors/library/constructors/constructors_version.js:9]: oops, then @Jesus89 I would have expected it to use the JS lib if it's generated

File: constructors/bq/sql/VERSION.sql
Comment (jgoizueta) [inline constructors/bq/sql/VERSION.sql:12]: Sorry for the noise, I misplace a comment about this: this is where I thing the JS version function should be used. (I'll take a look a view at the SETool PR, because I think it should handle this.

Comment (alex-graciano) [reply constructors/bq/sql/VERSION.sql:12]: Ok, no problem!

**Review by jgoizueta — CHANGES_REQUESTED** [review]
Regarding my comments about ST_MAKEPOLYGONORIENTED I leave it to you to decide, it's a very minor thing, the current implementation should be fine, but please fixe the version to use the JS function for future consistency.

File: constructors/README.md
Comment (vdelacruzb) [inline constructors/README.md:3]: Maybe you can add here the structure even if it's the default text. bq/ library/... you can grab it from other modules.

File: constructors/bq/test/README.md
Comment (vdelacruzb) [inline constructors/bq/test/README.md:3]: the tiler -> constructors

File: constructors/library/test/README.md
Comment (vdelacruzb) [inline constructors/library/test/README.md:3]: the tiler -> constructors.
This actually has to be updated in skel I think so we don't mess with it.

File: tools/setool/.gitignore
Comment (vdelacruzb) [inline tools/setool/.gitignore:2]: now that I see this... is package-lock autogenerated from package.json? If it's the case we can ignore it in the whole repository.

File: constructors/README.md
Comment (alex-graciano) [reply constructors/README.md:3]: Great, i'll update the PR accordingly.

**Review by jgoizueta — CHANGES_REQUESTED** [review]
I'm sorry to have brought in the question of LEAST/GREATEST which is probably not worth the time to make a decision os such small detail, but I leave that up to you.

But this time I think the order of the vertices must be reversed to be correctly interpreted by ST_MAKEPOLYGONORIENTED

**Comment by jgoizueta** [comment]
Hey, @alex-graciano you shouldn't be so quick to pay attention to my suggestions 😛 , I think I wasn't very inspired today and led you to spend time on something that was perfectly functional the first time, sorry about that.

**Comment by alex-graciano** [comment]
> Hey, @alex-graciano you shouldn't be so quick to pay attention to my suggestions 😛 , I think I wasn't very inspired today and led you to spend time on something that was perfectly functional the first time, sorry about that.

Don't worry! I'll take a look to the suggestions!

File: constructors/bq/sql/ST_MAKEENVELOPE.sql
Comment (alex-graciano) [reply constructors/bq/sql/ST_MAKEENVELOPE.sql:12]: WRT the ST_MAKEPOLYGONORIENTED I think now the orientation is now well established. At least it is in line now with the PostGIS implementation.

```sql
-- BigQuery
SELECT `cartodb-gcp-backend-data-team.agraciano_constructors`.ST_MAKEENVELOPE(10,10,11,11)
-- POLYGON((10 10, 10 11, 11 11, 11 10, 10 10))

-- PostGIS
SELECT ST_AsText(ST_MakeEnvelope(10,10,11,11))
-- POLYGON((10 10,10 11,11 11,11 10,10 10))
```

While if we use `ST_MAKEPOLYGON` in our implementation we obtain:

```sql
SELECT `cartodb-gcp-backend-data-team.agraciano_constructors`.ST_MAKEENVELOPE(10,10,11,11)
-- POLYGON((11 10, 11 11, 10 11, 10 10, 11 10))
```

Comment (jgoizueta) [reply constructors/bq/sql/ST_MAKEENVELOPE.sql:12]: Hey, after discussing this with @alex-graciano we agreed to leave the initial implementation where `ST_MAKEPOLYGON` will decide the orientation of the result (anti-clockwise as per BQ convention).

**Review by jgoizueta — APPROVED** [review]
🏁

**Comment by Jesus89** [comment]
Could you merge master to solve conflicts?

**Review by Jesus89 — APPROVED** [review]
Great work!

I'll wait for this to be deployed in staging and then merge to deploy to prod

**Comment by Jesus89** [comment]
We missed one point. Adding `BQ_DATASET_CONSTRUCTORS`. I'll fix it.

**Review by Jesus89 — APPROVED** [review]

File: constructors/Makefile
Comment (vdelacruzb) [inline constructors/Makefile:5]: this could be removed

File: constructors/Makefile
Comment (vdelacruzb) [inline constructors/Makefile:2]: It looks like something was going wrong in SF staging. Try removing "sf/" from here as this module does not contain sf implementation still

Comment (Jesus89) [reply constructors/Makefile:2]: Nope, this is because of the ENABLED_BQ=1 by default, it should be 0 and set by CI/CD scripts. Otherwise, the SF job tries to deploy BQ.

File: constructors/Makefile
Comment (Jesus89) [reply constructors/Makefile:5]: I have changed this to 0

**Review by Jesus89 — APPROVED** [review]

File: constructors/Makefile
Comment (vdelacruzb) [reply constructors/Makefile:2]: ooh true! ^^' . Then with your fix is good to go

**Review by vdelacruzb — APPROVED** [review]
Nicely done!

**Pull request merged**
